sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+                Computing the best log-probability creates a significant overhead
+                (thus, the default is `10`.)
+            show_progress_bars: Whether or not to show a progressbar for sampling from
+                the posterior.
+            individually: If true, returns log weights and MAPs individually.
+
+        Returns:
+            The ensemble MAP estimate or individual log_weigths and component MAP
+            estimate if individually == True.
+        """
+
+        if individually:
+            maps = []
+            log_weights = torch.log(self._weights).reshape(-1, 1)
+
+            for posterior in self.posteriors:
+                maps.append(
+                    posterior.map(
+                        x=x,
+                        num_iter=num_iter,
+                        num_to_optimize=num_to_optimize,
+                        learning_rate=learning_rate,
+                        init_method=init_method,
+                        num_init_samples=num_init_samples,
+                        save_best_every=save_best_every,
+                        show_progress_bars=show_progress_bars,
+                    )
+                )
+            maps = torch.stack(maps)
+            return log_weights, maps
+
+        else:
+            if init_method == "posterior":
+                inits = self.sample((num_init_samples,))
+            elif init_method == "proposal":
+                inits = self.proposal.sample((num_init_samples,))
+            elif isinstance(init_method, Tensor):
+                inits = init_method
+            else:
+                raise ValueError
+
+            return gradient_ascent(
+                potential_fn=self.potential_fn,
+                inits=inits,
+                theta_transform=self.theta_transform,
+                num_iter=num_iter,
+                num_to_optimize=num_to_optimize,
+                learning_rate=learning_rate,
+                save_best_every=save_best_every,
+                show_progress_bars=show_progress_bars,
+            )[0]
+
+
+class EnsemblePotentialProvider(BasePotential):
+    r"""Provides `NeuralPosteriorEnsemble` with a `potential_fn` attribute.
+
+    The potential is the same as the sum of the weighted log-probabilities of each
+    component posterior.
+
+    This class was modelled off of `PosteriorBasedPotential` and should provide similar
+    functionality.
+
+    Attributes:
+        posteriors: List of the posterior estimators making up the ensemble.
+        weights: Weight of each posterior distribution. If none are provided each
+            posterior is weighted with 1/N.
+        prior: Prior distribution that is the same for all posteriors. If it is not the
+            same, then `prior = None`.
+        device: Device which the component distributions sit on.
+        x_o: Used as conditioning context for `potential_fn`.
+    """
+    allow_iid_x = False  # type: ignore
```

Comment:
hm this depends. If the potential is based on `SNLE` or `SNRE`, it should be `True`. I would suggest setting it to True here. The error will then occur at `set_x()` in the individual posteriors. Please leave a comment here that this is what will happen

Reply:
I could also implement a similar check as in `log_prob`, i.e. `if areinstance(posteriors, SNPE): allow_iid_x = False else: allow_iid_x = True`

Reply:
that also works